typing/typecore.ml

```diff
@@ -670,92 +674,109 @@ end) = struct
         in
         List.find check_type lbls
 
+  (** [disambiguate] selects a concrete description for [lid] using some
+      contextual information: an optional [expected_type], and a list of candidates [lbls].
+      If [expected_type] is [None], it just returns the head of [lbls].
+      Otherwise, it extracts the description directly from the corresponding type definition,
+      except for extension types, where it has to choose it inside [lbls].
+      This code is careful to warn in a number of situations:
+      * if a non principal type was used for disambiguation [Not_principal]
+      * if [lbls] contained several valid candidates [Ambiguous_name]
+      * if the name was chosen out of [scope] [Name_out_of_scope]
+      Here [scope] contains the labels equal to [lid] in the current environment,
+      and [lbls] is a (potentially strict) subset of [scope], see [disambiguate_label_by_ids] *)
   let disambiguate ?(warn=Location.prerr_warning) ?scope
-                   usage lid env opath lbls =
+                   usage lid env expected_type lbls =
     let scope = match scope with None -> lbls | Some l -> l in
-    let lbl = match opath with
-      None ->
+    let warn_if_ambiguous lbl rest =
+      Printtyp.Conflicts.reset ();
+      let paths = ambiguous_types env lbl rest in
+      let expansion =
+        Format.asprintf "%t" Printtyp.Conflicts.print_explanations in
+      (* warn if there were multiple candidates in scope *)
+      if paths <> [] then
+        warn lid.loc
+          (Warnings.Ambiguous_name ([Longident.last lid.txt],
+                                    paths, false, expansion))
+    in
+    let warn_non_principal () =
+      let name = Datatype_kind.label_name kind in
+      warn lid.loc
+        (Warnings.Not_principal
+           ("this type-based " ^ name ^ " disambiguation"))
+    in
+    let warn_out_of_scope tpath =
+      let path_s =
+        Printtyp.wrap_printing_env ~error:true env
+          (fun () -> Printtyp.string_of_path tpath) in
+      warn lid.loc
+        (Warnings.Name_out_of_scope (path_s, [Longident.last lid.txt], false))
+    in
+    let warn_disambiguated_name lbl scope =
+      match scope with
+      | Ok ((lab1,_) :: _) when lab1 == lbl -> ()
+      | _ ->
+          Location.prerr_warning lid.loc
+            (Warnings.Disambiguated_name (get_name lbl))
+    in
```

Comment:
FTR: I refrained from saying so earlier, but I'm curious to see what the code would look like if these helpers were lifted out. Right now they are introducing 30 lines of noise between the code and its explanation.

Reply:
I pushed a commit that does it. It's not very nice because of the "warn" parameter -- it feels like code that was lambda-lifted by an automated tool rather than written by a human. I'm not convinced this is the best move, but let's wait to see how we like it.

Reply:
For once, I agree with @gasche: this code is about the warnings, which are only meaningful in that context, so moving them out does not improve things. Maybe we would be better off with a `where` syntax, but it was dropped in ocaml.

Reply:
(I think that @trefis' idea was also that sometimes, conciseness brings clarity, and the version with the warning code lifted out could result in a more brain-sized function that is easier to read and improve further.) I did think about using `let rec ... and ...` but I think it could be confusing.